crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1726 @@
+ */
+static void
+decrypt_cpa(uint8_t *out, const uint8_t *ctext, scalar *u,
+            const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+    scalar v, mask;
+    int rank = vinfo->rank;
+    int du = vinfo->du;
+    int dv = vinfo->dv;
+
+    vector_decode(u, ctext, du, rank);
+    vector_decompress(u, du, rank);
+    vector_ntt(u, rank);
+    scalar_decode(&v, ctext + vinfo->u_vector_bytes, dv);
+    scalar_decompress(&v, dv);
+    inner_product(&mask, key->s, u, rank);
+    scalar_inverse_ntt(&mask);
+    scalar_sub(&v, &mask);
+    scalar_compress(&v, 1);
+    scalar_encode_1(out, &v);
+}
+
+static void encode_pubkey(uint8_t *out, const ML_KEM_KEY *key)
+{
+    const uint8_t *rho = key->rho;
+    vinfo_t vinfo = key->vinfo;
+
+    vector_encode_12(out, key->t, vinfo->rank);
+    memcpy(out + vinfo->vector_bytes, rho, ML_KEM_RANDOM_BYTES);
+}
+
+static void encode_prvkey(uint8_t *out, const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    vector_encode_12(out, key->s, vinfo->rank);
+    out += vinfo->vector_bytes;
+    encode_pubkey(out, key);
+    out += vinfo->pubkey_bytes;
+    memcpy(out, key->pkhash, ML_KEM_PKHASH_BYTES);
+    out += ML_KEM_PKHASH_BYTES;
+    memcpy(out, key->z, ML_KEM_RANDOM_BYTES);
+}
+
+static int parse_pubkey(const uint8_t *in, EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    /* Decode and check |t| */
+    if (!vector_decode_12(key->t, in, vinfo->rank))
+        return 0;
+    /* Save the matrix |m| recovery seed |rho| */
+    memcpy(key->rho, in + vinfo->vector_bytes, ML_KEM_RANDOM_BYTES);
+    /*
+     * Pre-compute the public key hash, needed for both encap and decap.
+     * Also pre-compute the matrix expansion, stored with the public key.
+     */
+    return hash_h(key->pkhash, in, vinfo->pubkey_bytes, mdctx, key)
+        && matrix_expand(mdctx, key);
+}
+
+/* Loading of explicit private keys is a test-only interface. */
+static int parse_prvkey(const uint8_t *in, EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    /* Decode and check |s|. */
+    if (!vector_decode_12(key->s, in, vinfo->rank))
+        return 0;
+    in += vinfo->vector_bytes;
+
+    if (!parse_pubkey(in, mdctx, key))
+        return 0;
+    in += vinfo->pubkey_bytes;
+
+    /* Check public key hash. */
+    if (memcmp(key->pkhash, in, ML_KEM_PKHASH_BYTES) != 0)
+        return 0;
+    in += ML_KEM_PKHASH_BYTES;
+
+    memcpy(key->z, in, ML_KEM_RANDOM_BYTES);
+    return 1;
+}
+
+/*
+ * Key generation consumes a 32-byte RNG output plus 1 byte for the rank
+ * (domain separation) which are hashed together to produce a pair of
+ * 32-byte seeds public "rho" and private "sigma".
+ */
+static __owur
+int genkey(const uint8_t *seed, uint8_t *pubenc, scalar *tmp,
+           EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    uint8_t hashed[ML_KEM_SEED_BYTES];
+    const uint8_t *const sigma = hashed + ML_KEM_RANDOM_BYTES;
+    uint8_t augmented_seed[ML_KEM_RANDOM_BYTES + 1];
```

Comment:
should this be cleared at the end of the function? requires gotos instead of returns,

Reply:
I make some effort to clear heap-allocated sensitive storage, that might linger, but not stack-allocated sentive storage. It is nor realistically possible to wipe all sensitive data from the stack.
Keeping the stack cleaned means wiping all temporary copies of the `s` vector being encoded and decoded, various other intermediate results... :-(. So not going there.